pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,698 @@
+			},
+		},
+		pvUID: spec.PersistentVolume.UID,
+	}, nil
+}
+
+func (plugin *storageosPlugin) NewProvisioner(options volume.VolumeOptions) (volume.Provisioner, error) {
+	return plugin.newProvisionerInternal(options, &storageosUtil{})
+}
+
+func (plugin *storageosPlugin) newProvisionerInternal(options volume.VolumeOptions, manager storageosManager) (volume.Provisioner, error) {
+	return &storageosProvisioner{
+		storageosMounter: &storageosMounter{
+			storageos: &storageos{
+				manager: manager,
+				plugin:  plugin,
+			},
+		},
+		options: options,
+	}, nil
+}
+
+func (plugin *storageosPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volume.Spec, error) {
+	volNamespace, volName, err := getVolumeFromRef(volumeName)
+	if err != nil {
+		volNamespace = defaultNamespace
+		volName = volumeName
+	}
+	storageosVolume := &v1.Volume{
+		Name: volumeName,
+		VolumeSource: v1.VolumeSource{
+			StorageOS: &v1.StorageOSVolumeSource{
+				VolumeName:      volName,
+				VolumeNamespace: volNamespace,
+			},
+		},
+	}
+	return volume.NewSpecFromVolume(storageosVolume), nil
+}
+
+func (plugin *storageosPlugin) SupportsMountOption() bool {
+	return false
+}
+
+func (plugin *storageosPlugin) SupportsBulkVolumeVerification() bool {
+	return false
+}
+
+func getVolumeSource(spec *volume.Spec) (*v1.StorageOSVolumeSource, bool, error) {
+	if spec.Volume != nil && spec.Volume.StorageOS != nil {
+		return spec.Volume.StorageOS, spec.Volume.StorageOS.ReadOnly, nil
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.StorageOS != nil {
+		return spec.PersistentVolume.Spec.StorageOS, spec.ReadOnly, nil
+	}
+	return nil, false, fmt.Errorf("Spec does not reference a StorageOS volume type")
+}
+
+// storageosManager is the abstract interface to StorageOS volume ops.
+type storageosManager interface {
+	// Connects to the StorageOS API using the supplied configuration.
+	NewAPI(apiCfg *storageosAPIConfig) error
+	// Creates a StorageOS volume.
+	CreateVolume(provisioner *storageosProvisioner) (*storageosVolume, error)
+	// Attaches the disk to the kubelet's host machine.
+	AttachVolume(mounter *storageosMounter) (string, error)
+	// Detaches the disk from the kubelet's host machine.
+	DetachVolume(unmounter *storageosUnmounter, dir string) error
+	// Mounts the disk on the Kubelet's host machine.
+	MountVolume(mounter *storageosMounter, mnt, dir string) error
+	// Unmounts the disk from the Kubelet's host machine.
+	UnmountVolume(unounter *storageosUnmounter) error
+	// Deletes the storageos volume.  All data will be lost.
+	DeleteVolume(deleter *storageosDeleter) error
+}
+
+// storageos volumes represent a bare host directory mount of an StorageOS export.
+type storageos struct {
+	podUID       types.UID
+	podNamespace string
+	pvName       string
+	volName      string
+	volNamespace string
+	secretName   string
+	readOnly     bool
+	description  string
+	pool         string
+	fsType       string
+	size         int
```

Comment:
I see some other volume plugins name the field as sizeGB... probably a good idea